src/main/java/ru/ralsei/whatcanyousee/logic/MazeGameMap.java

```diff
@@ -0,0 +1,809 @@
+     */
+    Cell getRelatedCell(Coordinates cellCoordinates, Coordinates relatedCoordinates) {
+        return getCell(new Coordinates(cellCoordinates.getX() + relatedCoordinates.getX(), cellCoordinates.getY() + relatedCoordinates.getY()));
+    }
+
+    /**
+     * Returns cell with coordinates (cellX + dx, cellY + dy).
+     */
+    Cell getRelatedCell(Cell cell, Coordinates coordinates) {
+        return getCell(new Coordinates(cell.getX() + coordinates.getX(), cell.getY() + coordinates.getY()));
+    }
+
+    /**
+     * Returns if cell with given coordinates is wall, meaning you can't go through this cell.
+     */
+    boolean isWall(Coordinates coordinates) {
+        Cell cell = getCell(coordinates);
+        return cell != null && cell.isWall();
+    }
+
+    /**
+     * Returns if cell with given coordinates is trap, meaning stepping on this cell causing a special event.
+     */
+    boolean isTrap(Coordinates coordinates) {
+        Cell cell = getCell(coordinates);
+        return cell != null && cell.isTrap();
+    }
+
+    /**
+     * If player is staying on the trap, causing a special event of this trap.
+     */
+    void applyTrap() throws IllegalStateException {
+        if (!isTrap(currentCoordinates)) {
+            throw new IllegalStateException("Position with coordinates " + currentCoordinates.getX() + " " + currentCoordinates.getY() + " is not a trap");
+        }
+
+        getCurrentCell().getTrap().apply();
+    }
+
+    /**
+     * Returns true if cell with given coordinates is exit position.
+     */
+    boolean isExit(Coordinates coordinates) {
+        return exitCoordinates.equals(coordinates);
+    }
+
+    /**
+     * Returns true if conditions to exit has been fulfilled and player can win maze by just stepping on exit cell.
+     */
+    protected abstract boolean checkConditionToExit();
+
+    /**
+     * Trap is a special cell that causing a special event when you stepping on it.
+     * Basically, to create a trap one just has to implement that special event.
+     */
+    protected static abstract class Trap {
+        /**
+         * Event that happens when player steps on a trap.
+         */
+        protected abstract void apply();
+    }
+
+    /**
+     * Toogle is a special cell that causing a special event (opening door for instance)
+     * when player press button 'use' near toogle.
+     *
+     * Basically, to create a toogle one just has to implement that special event.
+     */
+    protected abstract class Toogle {
+        /**
+         * Special event cause by pressing this toogle.
+         */
+        protected abstract void use();
+    }
+
+    /**
+     * List of all monster in the game.
+     */
+    private final ArrayList<Monster> monsters = new ArrayList<>();
+
+    /**
+     * Adds new monster to the list.
+     */
+    protected void addMonster(Monster monster) {
+        monsters.add(monster);
+    }
+
+    public ArrayList<Monster> getMonsters() {
+        return monsters;
```

Comment:
This is not much different from just providing the public access to the monsters field. Can we return a copy instead, for encapsulation? Otherwise, I would again make the field public and remove the extra code associated with addMonster and getMonsters.

Reply:
Same as above, I can't force myself to make a public field (and we cannot return a copy).